xclim/sdba/base.py

```diff
@@ -606,3 +607,177 @@ def _apply_on_group(dsblock, **kwargs):
+      The coordinates of `movingwin` are the first date of the windows.
+    """
+    # Get number of samples per year (and perform checks)
+    N_in_year = _get_number_of_elements_by_year(da.time)
+
+    # Number of samples in a window
+    N = window * N_in_year
+
+    first_slice = da.isel(time=slice(0, N))
+    first_slice = first_slice.expand_dims({dim: np.atleast_1d(first_slice.time[0])})
+    daw = [first_slice]
+
+    i_start = N_in_year * step
+    # This is the first time I use `while` in real python code. What an event.
+    while i_start + N <= da.time.size:
+        # Cut and add _full_ slices only, partial window are thrown out
+        # Use isel so that we don't need to deal with a starting date.
+        slc = da.isel(time=slice(i_start, i_start + N))
+        slc = slc.expand_dims({dim: np.atleast_1d(slc.time[0])})
+        slc["time"] = first_slice.time
+        daw.append(slc)
+        i_start += N_in_year * step
+
+    daw = xr.concat(daw, dim)
+    return daw
+
+
+def unpack_moving_yearly_window(da: xr.DataArray, dim: str = "movingwin"):
+    """Unpack a constructed moving window dataset to a normal timeseries, only keeping the central data.
+
+    Unpack DataArrays created with :py:func:`construct_moving_yearly_window` and recreate a timeseries data.
+    Only keeps the central non-overlapping years. The final timeseries will be (window - step) years shorter than
+    the initial one.
+
+    The window length and window step are inferred from the coordinates.
+
+    Parameters
+    ----------
+    da: xr.DataArray
+      As constructed by :py:func:`construct_moving_yearly_window`.
+    dim : str
+      The window dimension name as given to the construction function.
+    """
+    # Get number of samples by year (and perform checks)
+    N_in_year = _get_number_of_elements_by_year(da.time)
+
+    # Might be smaller than the original moving window, doesn't matter
+    window = da.time.size / N_in_year
+
+    if window % 1 != 0:
+        warnings.warn(
+            f"Incomplete data received as number of years covered is not an integer ({window})"
+        )
+
+    # Get step in number of years
+    days_in_year = max_doy[get_calendar(da)]
+    step = np.unique(da[dim].diff(dim).dt.days / days_in_year)
+    if len(step) > 1:
+        raise ValueError("The spacing between the windows is not equal.")
+    step = int(step[0])
+
+    # Which years to keep: length step, in the middle of window
+    left = int((window - step) // 2)  # first year to keep
+
+    # Keep only the middle years
+    da = da.isel(time=slice(left * N_in_year, (left + step) * N_in_year))
+
+    out = []
+    for win_start in da[dim]:
+        slc = da.sel({dim: win_start}).drop_vars(dim)
+        dt = win_start.values - da[dim][0].values
+        slc["time"] = slc.time + dt
+        out.append(slc)
+
+    return xr.concat(out, "time")
+
+
+def stack_variables(ds):
```

Comment:
Is this something that should later be used in `spatial_analogs` ?

Reply:
True. This and `_escore` should be reused there.